firestore/src/include/firebase/firestore/filter.h

```diff
@@ -0,0 +1,96 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIREBASE_FIRESTORE_SRC_INCLUDE_FIREBASE_FIRESTORE_FILTER_H_
+#define FIREBASE_FIRESTORE_SRC_INCLUDE_FIREBASE_FIRESTORE_FILTER_H_
+
+#include <vector>
+
+#include "firebase/firestore/field_value.h"
+
+namespace firebase {
+namespace firestore {
+
+class FilterInternal;
+
+class Filter {
+ public:
+  static Filter ArrayContains(const std::string& field,
+                              const FieldValue& value);
+  static Filter ArrayContainsAny(const std::string& field,
+                                 const std::vector<FieldValue>& values);
+  static Filter EqualTo(const std::string& field, const FieldValue& value);
+  static Filter NotEqualTo(const std::string& field, const FieldValue& value);
+  static Filter GreaterThan(const std::string& field, const FieldValue& value);
+  static Filter GreaterThanOrEqualTo(const std::string& field,
+                                     const FieldValue& value);
+  static Filter LessThan(const std::string& field, const FieldValue& value);
+  static Filter LessThanOrEqualTo(const std::string& field,
+                                  const FieldValue& value);
+  static Filter In(const std::string& field,
+                   const std::vector<FieldValue>& values);
+  static Filter NotIn(const std::string& field,
+                      const std::vector<FieldValue>& values);
+
+  static Filter ArrayContains(const FieldPath& field, const FieldValue& value);
+  static Filter ArrayContainsAny(const FieldPath& field,
+                                 const std::vector<FieldValue>& values);
+  static Filter EqualTo(const FieldPath& field, const FieldValue& value);
+  static Filter NotEqualTo(const FieldPath& field, const FieldValue& value);
+  static Filter GreaterThan(const FieldPath& field, const FieldValue& value);
+  static Filter GreaterThanOrEqualTo(const FieldPath& field,
+                                     const FieldValue& value);
+  static Filter LessThan(const FieldPath& field, const FieldValue& value);
+  static Filter LessThanOrEqualTo(const FieldPath& field,
+                                  const FieldValue& value);
+  static Filter In(const FieldPath& field,
+                   const std::vector<FieldValue>& values);
+  static Filter NotIn(const FieldPath& field,
+                      const std::vector<FieldValue>& values);
+
+  template <class... Filters>
+  static Filter Or(const Filter& filter, const Filters&... filters);
```

Comment:
I just realized that the final decision was to allow empty filters (e.g. `Or()`) on other platforms. This came out of an API discussion where folks argued that when the developer is using a UI to build composite filters, if all filters are unselected, then the composite filter would contain no filters, and it would greatly simplify the developers' code if they don't have to guard against the corner case of "empty composite filter".

So other platforms (android/ios/web) currently allow empty Or and empty And filters. We should do the same thing here. (I realize the API proposal has to be updated as well).

So I think we'll have to go with 
```cpp
  template <class... Filters>
  static Filter Or(const Filters&... filters);
```
and change all relevant functions and constructors to adhere to this.